### PR TITLE
Bugfix/MADD(4)PG

### PR DESCRIPTION
**Pull request by DriesSmit**
## What?
Implements a few fixes and improvements to MADD(4)PG namely, #358, #357 and 2 more fixes.
## Why?
Bugfixes.
## How?
Changed the trainer files for MADDPG and MAD4PG.
## Extra
.

### CODE REVIEW

File: mava/systems/tf/maddpg/execution.py [hunk 1]
```diff
@@ -147,12 +147,19 @@ def select_action(
         # Step the recurrent policy/value network forward
         # given the current observation and state.
         action, policy = self._policy(agent, observation.observation)
-
-        # Return a numpy array with squeezed out batch dimension.
-        action = tf2_utils.to_numpy_squeeze(action)
-        policy = tf2_utils.to_numpy_squeeze(policy)
         return action, policy
 
+    @tf.function
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/execution.py:152]: @DriesSmit @RuanJohn we should just confirm this change didn't cause the memory leak issue.

File: mava/systems/tf/maddpg/execution.py [hunk 2]
```diff
@@ -367,9 +367,32 @@ def select_actions(
...
+        new_states = {}
         for agent, observation in observations.items():
-            actions[agent], policies[agent] = self.select_action(agent, observation)
-        return actions, policies
+            actions[agent], policies[agent], new_states[agent] = self.select_action(
+                agent, observation, states[agent]
+            )
+        return actions, policies, new_states
+
+    def select_action(  # type: ignore
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/execution.py:377]: Why do we have the type ignore here? Seems like basic typing.

File: mava/systems/tf/mad4pg/training.py [hunk 3]
```diff
@@ -594,7 +594,7 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
         data: Trajectory = inputs.data
 
         # Note (dries): The unused variable is start_of_episodes.
-        observations, actions, rewards, discounts, _, extras = (
+        observations, actions, rewards, end_of_episode, _, extras = (
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:597]: Much better naming :)

**Review by arnupretorius — COMMENTED** [review]
Thanks for the fixes @DriesSmit! 🙂  Just see my few minor comments.

File: mava/systems/tf/maddpg/execution.py (hunk 2 above)
Comment (DriesSmit) [reply mava/systems/tf/maddpg/execution.py:377]: I think it was not conforming to the recurrent executor's argument setup. We now use tf.function which means the agent_state should also be passed in as an argument. We could maybe update the recurrent executor definition when all the systems use tf.function over the select_actions function.

**Review by KaleabTessera — APPROVED** [review]
Thanks @DriesSmit :fire: :open_hands:

File: mava/systems/tf/maddpg/execution.py (hunk 2 above)
Comment (arnupretorius) [reply mava/systems/tf/maddpg/execution.py:377]: Ok thanks, perhaps we can make an issue for this.

File: mava/systems/tf/maddpg/execution.py (hunk 1 above)
Comment (arnupretorius) [reply mava/systems/tf/maddpg/execution.py:152]: Discussed offline. Still under investigation but is not particular to MADDPG.

**Review by arnupretorius — APPROVED** [review]

File: mava/systems/tf/maddpg/execution.py (hunk 2 above)
Comment (DriesSmit) [reply mava/systems/tf/maddpg/execution.py:377]: Will do.

**Pull request merged**